### PR TITLE
Implement hangfire framework

### PR DESCRIPTION
**Pull request by vldmrklnk**

### CODE REVIEW

File: InternshipBE/WebApi/appsettings.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
     "LogLevel": {
       "Default": "Information",
       "Microsoft": "Warning",
-      "Microsoft.Hosting.Lifetime": "Information"
+      "Microsoft.Hosting.Lifetime": "Information",
```
Comment (romanstat) [inline InternshipBE/WebApi/appsettings.json]: is it necessary?

File: InternshipBE/DAL/Repositories/CandidateRepository.cs [hunk 2]
```diff
@@ -18,7 +18,7 @@ public async Task<List<Candidate>> GetAllByInternshipIdAsync(int id, int itemsCo
 		{
 			var internship = await _context.Internships.Include(x => x.Candidate).FirstOrDefaultAsync(x => x.Id == id);
 
-			return internship?.Candidate.Skip(itemsCount * pageNumber).Take(itemsCount).ToList();
+			return internship?.Candidate.Skip(itemsCount * --pageNumber).Take(itemsCount).ToList();
```
Comment (romanstat) [inline InternshipBE/DAL/Repositories/CandidateRepository.cs]: This changes in another pull request

File: InternshipBE/WebApi/Startup.cs [hunk 3]
```diff
@@ -108,6 +116,7 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 			app.UseEndpoints(endpoints =>
 			{
 				endpoints.MapControllers();
+				endpoints.MapHangfireDashboard();
```
Comment (romanstat) [inline InternshipBE/WebApi/Startup.cs]: for what?

File: InternshipBE/WebApi/Startup.cs [hunk 4]
```diff
@@ -91,6 +96,9 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
 				app.UseDeveloperExceptionPage();
 				app.UseSwagger();
 				app.UseSwaggerUI(c => c.SwaggerEndpoint("/swagger/v1/swagger.json", "WebApi v1"));
+
+				app.UseHangfireDashboard();
+				RecurringJob.AddOrUpdate<IGoogleSheetService>("getnewcandidates", x => x.SaveNewCandidatesAsync(), Cron.Minutely);
```
Comment (romanstat) [inline InternshipBE/WebApi/Startup.cs]: for what first parametr - "getnewcandidates"?

File: InternshipBE/WebApi/Startup.cs (hunk 3 above)
Comment (RomanExadel) [reply InternshipBE/WebApi/Startup.cs:119]: or it is the same like app.UseHangfireServer(); app.UseHangfireDashboard();?

File: InternshipBE/WebApi/Startup.cs (hunk 4 above)
Comment (romanstat) [reply InternshipBE/WebApi/Startup.cs]: Cron.Minutely is it okey?

File: InternshipBE/BL/Mapping/Profiles/CandidateProfile.cs [hunk 5]
```diff
@@ -16,7 +16,7 @@ public CandidateProfile()
 				.ForMember(e => e.FirstName, source => source.MapFrom(s => s[FIRST_NAME_OFFSET].ToString()))
 				.ForMember(e => e.LastName, source => source.MapFrom(s => s[LAST_NAME_OFFSET].ToString()))
 				.ForMember(e => e.RegistationDate, source => source
-					.MapFrom(s => DateTime.ParseExact(s[REGISTRATION_DATE_OFFSET].ToString(), "dd.MM.yyyy HH:mm:ss", CultureInfo.InvariantCulture)))
+					.MapFrom(s => DateTime.ParseExact(s[REGISTRATION_DATE_OFFSET].ToString(), "dd.MM.yyyy H:mm:ss", CultureInfo.InvariantCulture)))
```
Comment (Amarozka123) [inline InternshipBE/BL/Mapping/Profiles/CandidateProfile.cs]: Move this string "dd.MM.yyyy H:mm:ss" to config

File: InternshipBE/WebApi/Startup.cs [hunk 6]
```diff
@@ -32,6 +34,12 @@ public void ConfigureServices(IServiceCollection services)
 		{
 			services.AddDbContext<ApplicationDbContext>(options => options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")));
 
+			services.AddHangfire(x =>
+					x.UseSqlServerStorage(Configuration.GetConnectionString("DefaultConnection"))
```
Comment (Amarozka123) [inline InternshipBE/WebApi/Startup.cs]: Create const string for "DefaultConnection"

File: InternshipBE/Shared/Config/CronConfiguration.cs [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+		{
+			var cronConfig = configuration["CronSetting"];
+			string format = null;
+
+			switch (cronConfig)
+			{
+				case "Minuteli":
+					format = Cron.Minutely();
+					break;
+				case "Hourly":
```
Comment (romanstat) [inline InternshipBE/Shared/Config/CronConfiguration.cs]: remove hardcode, use constants

File: InternshipBE/Shared/Config/CronConfiguration.cs [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+namespace Shared.Config
+{
+	public static class CronConfiguration
+	{
+		public static string ReturnCron(IConfiguration configuration)
+		{
+			var cronConfig = configuration["CronSetting"];
+			string format = null;
+
+			switch (cronConfig)
```
Comment (RomanExadel) [inline InternshipBE/Shared/Config/CronConfiguration.cs:13]: switch pattern matching

File: InternshipBE/Shared/Config/CronConfiguration.cs [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+			{
+				case "Minuteli":
+					format = Cron.Minutely();
+					break;
+				case "Hourly":
+					format = Cron.Hourly();
+					break;
+			}
+
+			return format;
```
Comment (romanstat) [inline InternshipBE/Shared/Config/CronConfiguration.cs]: switch pattern matching

File: InternshipBE/Shared/Config/Interfaces/IGoogleSheetConfig.cs [hunk 10]
```diff
@@ -11,5 +11,7 @@ public interface IGoogleSheetConfig
 		public string RangeSettings { get; }
 
 		public string ClientSecrets { get; }
+
+		public string DateTimeFormat{ get; }
```
Comment (romanstat) [inline InternshipBE/Shared/Config/Interfaces/IGoogleSheetConfig.cs:15]: aling

File: InternshipBE/WebApi/Controllers/HangfireController.cs [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+	{
+		private readonly IGoogleSheetService _googleSheetService;
+
+		public HangfireController(IGoogleSheetService googleSheetService)
+		{
+			_googleSheetService = googleSheetService;
+		}
+
+		[HttpGet]
+		public async Task<IActionResult> UpdateCandidatesAsync()
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/HangfireController.cs]: remove async in name

File: InternshipBE/WebApi/Startup.cs [hunk 12]
```diff
@@ -30,7 +34,13 @@ public Startup(IConfiguration configuration)
 		// This method gets called by the runtime. Use this method to add services to the container.
 		public void ConfigureServices(IServiceCollection services)
 		{
-			services.AddDbContext<ApplicationDbContext>(options => options.UseSqlServer(Configuration.GetConnectionString("DefaultConnection")));
+			services.AddDbContext<ApplicationDbContext>(options => options.UseSqlServer(Configuration.GetConnectionString(connectionString)));
+
+			services.AddHangfire(x =>
```
Comment (romanstat) [inline InternshipBE/WebApi/Startup.cs]: in one line

File: InternshipBE/WebApi/Controllers/GoogleSheetController.cs [hunk 13]
```diff
@@ -0,0 +1,26 @@
+﻿using BL.Interfaces;
+using Microsoft.AspNetCore.Mvc;
+using System.Threading.Tasks;
+
+namespace WebApi.Controllers
+{
+	[Route("api/hangfire")]
```
Comment (romanstat) [inline InternshipBE/WebApi/Controllers/GoogleSheetController.cs]: api/[controller]

**Review by romanstat — APPROVED** [review]

**Review by Xodoc — APPROVED** [review]

**Review by Evgeniy-Sinicin — APPROVED** [review]

**Pull request merged**